File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.util.Scanner;


public class Main {
	
	public static void main(String[] args){
    	Scanner sc = new Scanner(System.in);
    	
    	int n = sc.nextInt(); 
    	int[] lst = new int[n + 1];
    	int[] dp = new int[n + 1];
    	
    	for(int i = 1;i < n + 1;i++) {
    		lst[i] = sc.nextInt();
    	}
    	dp[1] = lst[1];
    	if (n > 1)
    		dp[2] = lst[1] + lst[2];

    	for(int i = 3;i < n + 1;i++) {
    		dp[i] = Math.max(dp[i - 3] + lst[i] + lst[i - 1], dp[i - 2] + lst[i]);
    		dp[i] = Math.max(dp[i], dp[i - 1]);
    	}
    	System.out.println(dp[n]);
	}
}

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
n = int(input())
lst = [int(input()) for _ in range(n)]

dp = [0] * n

dp[0] = lst[0]
if n > 1:
    dp[1] = lst[1] + lst[0]
if n > 2:
    dp[2] = max(lst[1] + lst[2], lst[2] + lst[0], dp[1])

for i in range(3, n):
    dp[i] = max(dp[i - 3] + lst[i] + lst[i - 1], dp[i - 2] + lst[i], dp[i - 1])
print(dp[n - 1])